File: 二叉树/二叉搜索树中第K小的元素.java
package 二叉树;

import java.util.ArrayList;
import java.util.List;

public class 二叉搜索树中第K小的元素 {
    int K;
    List<Integer> nums = new ArrayList<Integer>();

    public int kthSmallest(TreeNode root, int k) {
        K = k;
        help(root);
        return nums.get(nums.size() - 1);

    }

    public void help(TreeNode root) {
        if (root == null) return;
        help(root.left);
        //中序遍历到第k个结束
        if (nums.size() == K) {
            return;
        }
        nums.add(root.val);
        help(root.right);
    }
}

class ewfcwe {
    int num = 0;
    int res;
    int K;

    public int kthSmallest(TreeNode root, int k) {
        K = k;
        dfs(root);
        return res;
    }

    public void dfs(TreeNode root) {
        if (root == null) return;
        dfs(root.left);
        num++;
        //目前所在的节点是第num个节点
        if (num == K) {
            res = root.val;
            return;
        }
        dfs(root.right);
    }
}

File: 二叉树/二叉搜索树中第k小的元素.go
package 二叉树

func kthSmallest(root *TreeNode, k int) int {
	var arr []int
	var inorder func(root *TreeNode)
	inorder = func(root *TreeNode) {
		if root == nil {
			return
		}
		//if len(arr) == k { //中序遍历到第k个停止
		//	return
		//}
		inorder(root.Left)
		if len(arr) == k { //中序遍历到第k个停止  思考一下为什么把截止放在这里呢 ，截止操作放在数组添加元素之前
			return
		}
		arr = append(arr, root.Val)
		inorder(root.Right)
	}
	inorder(root)
	return arr[len(arr)-1]
}
